### PR TITLE
Potential fix for bug in "Use These Parameters" popup menu entry.

### PR DESCRIPTION
**Pull request by msmttchr**
The fix is actually changing the set_parameter API and it is setting the current value instead of the default value.

### CODE REVIEW

**Comment by msmttchr** [comment]
Indeed this pull request is very similar to #330. The solution is slightly different and the test issue is similar. In #330, I think there is some redundant check, but I am not sure.

**Comment by CasperSchippers** [comment]
It seems indeed that your solution works as intended and is shorter than #330.
I will close #330.

**Review by CasperSchippers — CHANGES_REQUESTED** [review]
I've had a look at this unit-test error. For some reason `parameter.value` may only be called once in the test (i.e. not once in the `input.set_parameter` method, as it is also called in the test in `input.update_parameter`, as a part of the call to `bool_input.parameter`). Why this is specifically not allowed is not completely clear to me. Maybe @cjermain or @bilderbuchi knows why this is so.

Either case, it can be fixed by slightly adjusting the solution in `set_parameter`.

Moreover, could you add yourself to `AUTHORS.txt` and `docs/about/authors.rst`.

File: pymeasure/display/inputs.py
Comment (msmttchr) [reply pymeasure/display/inputs.py:58]: I agree using is_set is more appropriate in my opinion.

File: pymeasure/display/inputs.py
Comment (msmttchr) [reply pymeasure/display/inputs.py:59]: Hello Casper,

first of all, I agree with You that this is a workaround to please the test more than a proper solution.
I did some research on python unittest module and I have tried to figure out the rationale behind the test.
The test is trying to assess that when you use the BooleanInput widget and change its value, the property 'value' associated is updated exactly once and with the expected value.
Since during the class instantiation ("BooleanInput(bool_param)"), the 'value' property is read twice, this is accounted as two calls with no parameters in the mock.
My proposal is to add the following code to fix the test (see line "p.reset_mock()"):

    def test_leftclick_should_update_parameter(self, qtbot, default_value):
        # set up BooleanInput

        bool_param = BooleanParameter(potato, default=default_value)
        with mock.patch('test_inputs.BooleanParameter.value',
                new_callable=mock.PropertyMock,
                return_value=default_value) as p:
            bool_input = BooleanInput(bool_param)

            # Clear any call to property 'value' during initialization
            p.reset_mock()

            qtbot.addWidget(bool_input)
            bool_input.show()

            # TODO: fix: fails to toggle on Windows
            #qtbot.mouseClick(bool_input, QtCore.Qt.LeftButton)
            bool_input.setValue(not default_value)

            assert bool_input.value() == (not default_value)
            bool_input.parameter # lazy update
            p.assert_called_once_with(not default_value)

Does this make sense to You ?

Michele

Comment (bilderbuchi) [reply pymeasure/display/inputs.py:59]: I agree that we should not change the implementation to "please the test".
As to why the tests asserts on one call, maybe this was just an oversight in the test design shadowed by the bug you fix in here. @Laogeodritt implemented this test, maybe you can comment?
The proposed change to the test sounds reasonable to me. Great that you already thought of a comment explaining why that line is in there.

**Comment by bilderbuchi** [comment]
What I would love to see in this PR is a test that shows that the observed behaviour was wrong before. 
As you already implemented a fix, make sure to check that the test fails without your fix (e.g. by temporarily undoing the fix). 
Then, activate your fix an confirm that it makes the test succeed. 
See e.g. [here](https://en.wikipedia.org/wiki/Test-driven_development#Test-driven_development_cycle) for some more info on test-driven-development. Especially for bug fixes this can be very useful: this way, we have another test making sure that we don't regress to unwanted behaviour in the future. :-) 

From #330, I see that the bug statement is
> When using the option "Use these values" on an entry in the BrowserWidget, the default values for an input are inserted into the input, rather than the actual values used in that entry.

This should aid you in constructing an appropriate test, we are happy to assist in case you run into problems!

File: pymeasure/display/inputs.py
Comment (CasperSchippers) [reply pymeasure/display/inputs.py:59]: Thanks for diving in Michele! I agree that the proposed fix in the test sounds reasonable and clear.

**Comment by msmttchr** [comment]
> What I would love to see in this PR is a test that shows that the observed behaviour was wrong before.
> As you already implemented a fix, make sure to check that the test fails without your fix (e.g. by temporarily undoing the fix).
> Then, activate your fix an confirm that it makes the test succeed.
> See e.g. [here](https://en.wikipedia.org/wiki/Test-driven_development#Test-driven_development_cycle) for some more info on test-driven-development. Especially for bug fixes this can be very useful: this way, we have another test making sure that we don't regress to unwanted behaviour in the future. :-)
> 
> From #330, I see that the bug statement is
> 
> > When using the option "Use these values" on an entry in the BrowserWidget, the default values for an input are inserted into the input, rather than the actual values used in that entry.
> 
> This should aid you in constructing an appropriate test, we are happy to assist in case you run into problems!
@bilderbuchi , I have submitted an update of the pull requests which adds the tests. In doing so, I had to change also the ListInput class, but I believe changes are ok. Please have a look if the test style is OK.

**Review by CasperSchippers — APPROVED** [review]
This looks good to me. Test are holding up as you would expect/hope. Good job!!

As I do not have much experience with the testing part jet, I'd like @bilderbuchi to have final check before merging.

File: tests/display/test_inputs.py
Comment (bilderbuchi) [inline tests/display/test_inputs.py:45]: Thank you for adding/adapting tests for this!

(This applies to all the tests changed/added here)
You are basically duplicating the test code here, as your comment also hints at a little. In general it's advisable to limit the number of asserts and different logic/configurations in any given test. This limits the opportunities for a test to break, which eases diagnosis _if_ the test fails.

Two approaches to improve the situation are:
* To split this into two tests with different names to indicate the two different "purposes" for the tests. This can duplicate test code of course, which is typically not very desirable but can be worth it if the differences between the test implementations are larger.
* Another way, especially if the code of the two different alternatives is nearly identical, is to stack another `parametrize` along the different option set you choose (see the end of [this doc section](https://docs.pytest.org/en/stable/parametrize.html#pytest-mark-parametrize-parametrizing-test-functions)), which can be an elegant way to generate lots of combinations.

In this case, I think this should do the trick (decorating the test function, not the class, so only this test is affected)
```
@pytest.mark.parametrize("default_value", [True, False])
class TestBooleanInput:
    @pytest.mark.parametrize("value_remains_default", [True, False])
    def test_init_from_param(self, qtbot, default_value, value_remains_default):
``` 
then in the test do something like
```
if value_remains_default:
    ref = bool_param.default
else:
    ref = not bool_param.default
```
This should generate 4 tests with the approprite combinations without any duplicated test code. :-)

File: tests/display/test_inputs.py
Comment (bilderbuchi) [inline tests/display/test_inputs.py:98]: as above, adding a `parametrize` like ```@pytest.mark.parametrize("reference", ['default', 'second_choice'])``` and adapting the logic should do the trick.

File: tests/display/test_inputs.py
Comment (bilderbuchi) [inline tests/display/test_inputs.py:167]: as above, try adding a parametrize like @pytest.mark.parametrize("reference", ['min_', 'default']) and using `locals()[reference]` to access the appropriate value should work as expected.

**Review by bilderbuchi — CHANGES_REQUESTED** [review]
Thank you for the additional changes, I left a couple of questions and suggestions for improving the code.

File: tests/display/test_inputs.py
Comment (msmttchr) [reply tests/display/test_inputs.py]: Thanks you for the hints. I think what you proposed is very elegant and make a lot of sense. I have added in the update of the pull request.

File: pymeasure/display/inputs.py
Comment (msmttchr) [reply pymeasure/display/inputs.py]: Well, I tried to make as little changes as possible, but in general I agree with your comment and I have removed the new method _addSuffix

File: pymeasure/display/inputs.py
Comment (msmttchr) [reply pymeasure/display/inputs.py:216]: I think the call of the setSuffix method is needed for the input widgets of class QSpinBox or QDoubleSpinBox (both they a setSuffix method) and it is needed to set the unit.
Since the other widget did not have this setSuffix method, the original coder implemented them as dummy calls, I guess.

Comment (bilderbuchi) [reply pymeasure/display/inputs.py:216]: Yeah, if the intent was not to actively disable/no-op the setsuffix methods of the Qt superclass, I think we should be able to remove them now that all our implementations only have `pass`. @feinsteinben I think you touched that code last, any thoughts/insight on this?

Comment (CasperSchippers) [reply pymeasure/display/inputs.py:216]: There does seem to be a case where the `setSuffix` method has to be implemented: whenever the `VectorParameter` or the `PhysicalParameter` (which, by the way, is not imported in `experiment.__init__.py`, unlike the other Parameters) are used, `units` is a valid attribute and therefore `setSuffix` will be called in `Input.set_parameter`.
And, as @msmttchr mentions, the call to `setSuffix` in `Input.set_parameter` is indeed required for correctly setting the unit of the parameter.

Comment (bilderbuchi) [reply pymeasure/display/inputs.py:216]: OK, but now we have 3 empty `setSuffix` implementations, so how should we best correct that? It seems the logic around different combinations of Parameter subclasses and widget types, and which ones need suffixes when, is a bit unclear/gnarly - can we simplify that? I'll defer this decision to you (or Colin), as I am not really familiar with this part of the codebase.

Comment (CasperSchippers) [reply pymeasure/display/inputs.py:216]: The best way that I can think of to make it more clear/convenient is to make sure there is a dedicated input for each type of in inputs (right now, only for the VectorParameter and the PhysicalParameter, which is derived from the VectorParameter there are no dedicated inputs).
As this is something that I think would fall a bit outside the scope of this PR, I think that for now there are a couple of options for solving the `pass` implementation: leave it be for now, add an additional check (`hasattr(self, 'setSuffix')`) to the if statement around the call to `setSuffix` in `set_parameter`, or move the call to `setSuffix` to a re-implementation of the `set_parameter` method in the specific sub-classes of `Input` that actually needs this call.
Although this last option seems to me rather nice (as you don't refer to a method that the class might not have), it results in a bit of duplicate code (though it would be only a single line in the `ScientificInput`, `FloatInput`, and `IntegerInput`).

Comment (msmttchr) [reply pymeasure/display/inputs.py:216]: My two cents: I would leave it as it is for now, since some more restructuring could happen in the future and solve the problem.

Comment (bilderbuchi) [reply pymeasure/display/inputs.py:216]: ok then, I created #365 to keep track of this, further inputs welcome over there.

**Pull request merged**
